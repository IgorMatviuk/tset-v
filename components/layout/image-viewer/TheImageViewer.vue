<template>
  <transition name="fade">
    <div class="image-viewer">
      <div
        v-swiper:imageViewerSlider="swiperOptions"
        :cleanup-styles-on-destroy="false"
        class="image-viewer__slider"
      >
        <div class="swiper-wrapper image-viewer__slider-wrapper">
          <div
            v-for="(item, index) in items"
            :key="index"
            class="swiper-slide image-viewer__slide"
          >
            <div class="image-viewer__image-wrapper">
              <picture>
                <source
                  :data-srcset="item.preview_lg | webp"
                  srcset="/img/stub.png"
                  media="(max-width: 1439px)"
                  type="image/webp"
                />
                <source
                  :data-srcset="item.preview_lg"
                  srcset="/img/stub.png"
                  media="(max-width: 1439px)"
                  type="image/jpeg"
                />
                <source
                  :data-srcset="item.preview_xl | webp"
                  srcset="/img/stub.png"
                  type="image/webp"
                />
                <img
                  :data-src="item.preview_xl"
                  :title="item.title"
                  :alt="item.alt"
                  src="/img/stub.png"
                  class="image-viewer__image lazyload"
                  loading="lazy"
                  data-object-fit="cover"
                />
              </picture>
            </div>
          </div>
        </div>
        <SliderArrow arrow-type="prev" class="image-viewer__prev" />
        <SliderArrow arrow-type="next" class="image-viewer__next" />
        <button class="image-viewer__close" type="button" @click="closeGallery">
          <svg-icon
            name="close"
            class="image-viewer__icon"
            width="12"
            height="12"
          />
        </button>
      </div>
    </div>
  </transition>
</template>

<script>
import SliderArrow from '@/components/shared/SliderArrow'

import { mutationTypes as productMutationsTypes } from '@/store/product'

import imageNextGen from '@/mixins/image-next-gen'

export default {
  components: {
    SliderArrow
  },
  mixins: [imageNextGen],
  props: {
    activeIndex: {
      type: Number,
      required: true
    }
  },
  data() {
    return {
      swiperOptions: {
        watchOverflow: true,
        effect: 'fade',
        initialSlide: this.activeIndex,
        navigation: {
          nextEl: '.image-viewer__next',
          prevEl: '.image-viewer__prev'
        },
        keyboard: {
          enabled: true,
          onlyInViewport: false
        },
        breakpoints: {
          1200: {
            allowTouchMove: false
          }
        }
      }
    }
  },
  computed: {
    items() {
      const formattedPreviews = this.$store.state.product.activeVariant.previews.filter(
        (preview) => preview.preview_type !== 'video'
      )

      return formattedPreviews
    }
  },
  mounted() {
    window.addEventListener('keydown', this.closeOnEscapePress)
  },
  beforeDestroy() {
    window.removeEventListener('keydown', this.closeOnEscapePress)
  },
  methods: {
    closeGallery() {
      this.$store.commit(
        `product/${productMutationsTypes.SET_IMAGE_VIEWER_VISIBILITY_TOGGLE}`,
        false
      )
    },
    closeOnEscapePress($event) {
      /* istanbul ignore else */
      if ($event.code === 'Escape') {
        this.closeGallery()
      }
    }
  }
}
</script>

<style lang="scss">
.image-viewer {
  position: fixed;
  top: 0;
  right: 0;
  bottom: 0;
  left: 0;
  z-index: 999;
  // display: none;
  width: 100%;
  height: 100%;
  background-color: $white;

  &__slider {
    height: 100%;
    padding-top: 10px;
    padding-bottom: 10px;
  }

  &__wrapper {
    height: 100%;
  }

  &__slide {
    display: flex;
    align-items: center;
    justify-content: center;
    height: 100%;
  }

  &__image-wrapper {
    height: 100%;
  }

  &__image {
    display: block;
    max-width: 100%;
    max-height: 100%;
  }

  &__next,
  &__prev {
    position: absolute;
    top: 50%;
    color: $black;
    background-color: $white;
    border-radius: 50%;
    transform: translateY(-50%);
  }

  &__next {
    right: 15px;

    &.swiper-button-next {
      width: 50px;
      height: 50px;
    }
  }

  &__prev {
    left: 15px;

    &.swiper-button-prev {
      width: 50px;
      height: 50px;
    }
  }

  &__close {
    @include reset-button-styles();

    position: absolute;
    top: 15px;
    right: 15px;
    z-index: 2;
    display: flex;
    align-items: center;
    justify-content: center;
    width: 50px;
    height: 50px;
    color: $black;
    background-color: $white;
    border-radius: 50%;
  }

  &__icon {
    width: 20px;
    height: 20px;
  }

  @media only screen and (max-width: 1199.98px) and (orientation: portrait) {
    &__slider {
      padding: 0;
    }

    &__image {
      width: 100%;
      height: 100%;
      object-fit: cover;
      object-position: center center;
    }
  }

  @include media-breakpoint-down(sm) {
    &__next {
      &.swiper-button-next {
        width: 35px;
        height: 35px;
      }

      .icon {
        width: 19px;
        height: 6px;
      }
    }

    &__prev {
      &.swiper-button-prev {
        width: 35px;
        height: 35px;
      }

      .icon {
        width: 19px;
        height: 6px;
      }
    }

    &__close {
      top: 12px;
      right: 12px;
      width: 35px;
      height: 35px;
    }

    &__icon {
      width: 12px;
      height: 12px;
    }
  }
}
</style>
