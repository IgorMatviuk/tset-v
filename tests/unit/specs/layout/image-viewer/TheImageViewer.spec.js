import { shallowMount, createLocalVue } from '@vue/test-utils'
import VueAwesomeSwiper from 'vue-awesome-swiper'
import Vuex from 'vuex'

import _cloneDeep from 'lodash.clonedeep'

import TheImageViewer from '@/components/layout/image-viewer/TheImageViewer'

import {
  mutations as productMutations,
  getters as productGetters
} from '@/store/product'

import createFactory from '@/api/helpers/factory'
import productAttrs from '@/api/factories/product'

const localVue = createLocalVue()
localVue.use(VueAwesomeSwiper)
localVue.use(Vuex)

describe('TheImageViewer', () => {
  let store
  let activeVariant
  let activePreview

  function createTheImageViewer(props) {
    const wrapper = shallowMount(TheImageViewer, {
      localVue,
      store,
      propsData: {
        ...props
      },
      stubs: ['svg-icon']
    })

    return wrapper
  }

  beforeEach(() => {
    const products = createFactory(productAttrs, 12)

    activeVariant = products[0].variants[0]
    activePreview = activeVariant.previews[0]

    store = new Vuex.Store({
      modules: {
        product: {
          namespaced: true,
          state: {
            activeVariant: _cloneDeep(activeVariant),
            imageViewerVisibility: true,
            activePreview: _cloneDeep(activePreview)
          },
          mutations: { ...productMutations },
          getters: { ...productGetters }
        }
      }
    })
  })

  test('create image gallery', () => {
    const wrapper = createTheImageViewer({ activeIndex: 1 })

    expect(wrapper.element).toMatchSnapshot()
  })

  test('close image gallery', () => {
    const wrapper = createTheImageViewer({ activeIndex: 1 })

    const closeButton = wrapper.find('.image-viewer__close')
    closeButton.trigger('click')
    expect(wrapper.vm.$store.state.product.imageViewerVisibility).toBe(false)
  })

  test('close image gallery by clicking on esc', () => {
    const wrapper = createTheImageViewer({ activeIndex: 1 })

    const escapeEvent = new KeyboardEvent('keydown', { code: 'Escape' })
    window.dispatchEvent(escapeEvent)
    expect(wrapper.vm.$store.state.product.imageViewerVisibility).toBe(false)
  })

  test('destroy image gallery', () => {
    const wrapper = createTheImageViewer({ activeIndex: 1 })

    wrapper.destroy()
  })
})
