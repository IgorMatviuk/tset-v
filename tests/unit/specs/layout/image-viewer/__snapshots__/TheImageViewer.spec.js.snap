// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TheImageViewer create image gallery 1`] = `
<transition-stub
  name="fade"
>
  <div
    class="image-viewer"
  >
    <div
      class="image-viewer__slider swiper-container swiper-container-fade swiper-container-initialized swiper-container-horizontal"
    >
      <div
        class="swiper-wrapper image-viewer__slider-wrapper"
        style="webkit-transition-duration: 0ms; transition-duration: 0ms;"
      >
        <div
          class="swiper-slide image-viewer__slide"
        >
          <div
            class="image-viewer__image-wrapper"
          >
            <picture>
              <source
                data-srcset="/img/product/product__medium-1.webp"
                media="(max-width: 1439px)"
                srcset="/img/stub.png"
                type="image/webp"
              />
               
              <source
                data-srcset="/img/product/product__medium-1.jpg"
                media="(max-width: 1439px)"
                srcset="/img/stub.png"
                type="image/jpeg"
              />
               
              <source
                data-srcset="/img/product/product__large-1.webp"
                srcset="/img/stub.png"
                type="image/webp"
              />
               
              <img
                class="image-viewer__image lazyload"
                data-object-fit="cover"
                data-src="/img/product/product__large-1.jpg"
                loading="lazy"
                src="/img/stub.png"
                title="some title"
              />
            </picture>
          </div>
        </div>
        <div
          class="swiper-slide image-viewer__slide"
        >
          <div
            class="image-viewer__image-wrapper"
          >
            <picture>
              <source
                data-srcset="/img/product/product__medium-2.webp"
                media="(max-width: 1439px)"
                srcset="/img/stub.png"
                type="image/webp"
              />
               
              <source
                data-srcset="/img/product/product__medium-2.jpg"
                media="(max-width: 1439px)"
                srcset="/img/stub.png"
                type="image/jpeg"
              />
               
              <source
                data-srcset="/img/product/product__large-2.webp"
                srcset="/img/stub.png"
                type="image/webp"
              />
               
              <img
                class="image-viewer__image lazyload"
                data-object-fit="cover"
                data-src="/img/product/product__large-2.jpg"
                loading="lazy"
                src="/img/stub.png"
                title="some title"
              />
            </picture>
          </div>
        </div>
        <div
          class="swiper-slide image-viewer__slide"
        >
          <div
            class="image-viewer__image-wrapper"
          >
            <picture>
              <source
                data-srcset="/img/product/product__medium-3.webp"
                media="(max-width: 1439px)"
                srcset="/img/stub.png"
                type="image/webp"
              />
               
              <source
                data-srcset="/img/product/product__medium-3.jpg"
                media="(max-width: 1439px)"
                srcset="/img/stub.png"
                type="image/jpeg"
              />
               
              <source
                data-srcset="/img/product/product__large-3.webp"
                srcset="/img/stub.png"
                type="image/webp"
              />
               
              <img
                class="image-viewer__image lazyload"
                data-object-fit="cover"
                data-src="/img/product/product__large-3.jpg"
                loading="lazy"
                src="/img/stub.png"
                title="some title"
              />
            </picture>
          </div>
        </div>
        <div
          class="swiper-slide image-viewer__slide"
        >
          <div
            class="image-viewer__image-wrapper"
          >
            <picture>
              <source
                data-srcset="/img/product/product__medium-4.webp"
                media="(max-width: 1439px)"
                srcset="/img/stub.png"
                type="image/webp"
              />
               
              <source
                data-srcset="/img/product/product__medium-4.jpg"
                media="(max-width: 1439px)"
                srcset="/img/stub.png"
                type="image/jpeg"
              />
               
              <source
                data-srcset="/img/product/product__large-4.webp"
                srcset="/img/stub.png"
                type="image/webp"
              />
               
              <img
                class="image-viewer__image lazyload"
                data-object-fit="cover"
                data-src="/img/product/product__large-4.jpg"
                loading="lazy"
                src="/img/stub.png"
                title="some title"
              />
            </picture>
          </div>
        </div>
        <div
          class="swiper-slide image-viewer__slide"
        >
          <div
            class="image-viewer__image-wrapper"
          >
            <picture>
              <source
                data-srcset="/img/product/product__medium-1.webp"
                media="(max-width: 1439px)"
                srcset="/img/stub.png"
                type="image/webp"
              />
               
              <source
                data-srcset="/img/product/product__medium-1.jpg"
                media="(max-width: 1439px)"
                srcset="/img/stub.png"
                type="image/jpeg"
              />
               
              <source
                data-srcset="/img/product/product__large-1.webp"
                srcset="/img/stub.png"
                type="image/webp"
              />
               
              <img
                class="image-viewer__image lazyload"
                data-object-fit="cover"
                data-src="/img/product/product__large-1.jpg"
                loading="lazy"
                src="/img/stub.png"
                title="some title"
              />
            </picture>
          </div>
        </div>
      </div>
       
      <sliderarrow-stub
        arrowtype="prev"
        class="image-viewer__prev"
      />
       
      <sliderarrow-stub
        arrowtype="next"
        class="image-viewer__next"
      />
       
      <button
        class="image-viewer__close"
        type="button"
      >
        <svg-icon-stub
          class="image-viewer__icon"
          height="12"
          name="close"
          width="12"
        />
      </button>
      <span
        aria-atomic="true"
        aria-live="assertive"
        class="swiper-notification"
      />
    </div>
  </div>
</transition-stub>
`;
